src/engine/math_util.c

```diff
@@ -697,91 +786,70 @@ void mtxf_align_terrain_triangle(Mat4 mtx, Vec3f pos, s32 yaw, f32 radius) {
+        ap += 4;
+        destp++;
     }
-    vec3f_add(&dest[3][0], &b[3][0]);
+    vec3f_add(dest[3], b[3]);
     ((u32 *) dest)[15] = FLOAT_ONE;
 }
 
-/**
- * Set matrix 'dest' to 'mtx' scaled by vector s
- */
-void mtxf_scale_vec3f(Mat4 dest, Mat4 mtx, register Vec3f s) {
-    register f32 *temp  = (f32 *)dest;
-    register f32 *temp2 = (f32 *)mtx;
-    register s32 i;
-
-    for (i = 0; i < 4; i++) {
-        temp[ 0] = temp2[ 0] * s[0];
-        temp[ 4] = temp2[ 4] * s[1];
-        temp[ 8] = temp2[ 8] * s[2];
-        temp[12] =  temp2[12];
-        temp++;
-        temp2++;
-    }
-}
-
-/**
- * Multiply a vector with a transformation matrix, which applies the transformation
- * to the point. Note that the bottom row is assumed to be [0, 0, 0, 1], which is
- * true for transformation matrices if the translation has a w component of 1.
- */
-UNUSED void mtxf_mul_vec3s(Mat4 mtx, Vec3s b) {
-    register f32 x = b[0];
-    register f32 y = b[1];
-    register f32 z = b[2];
-    register f32 *temp2 = (f32 *)mtx;
-    register s32 i;
-    register s16 *c = b;
-    for (i = 0; i < 3; i++) {
-        c[0] = ((x * temp2[ 0])
-              + (y * temp2[ 4])
-              + (z * temp2[ 8])
-              +      temp2[12]);
-        c++;
-        temp2++;
-    }
-}
-
 /**
  * Set 'mtx' to a transformation matrix that rotates around the z axis.
  */
 #define MATENTRY(a, b)                          \
     ((s16 *) mtx)[a     ] = (((s32) b) >> 16);  \
     ((s16 *) mtx)[a + 16] = (((s32) b) & 0xFFFF);
 void mtxf_rotate_xy(Mtx *mtx, s32 angle) {
-    register s32 i = (coss(angle) * 0x10000);
-    register s32 j = (sins(angle) * 0x10000);
-    register f32 *temp = (f32 *)mtx;
-    register s32 k;
-    for (k = 0; k < 16; k++) {
-        *temp = 0;
-        temp++;
+    s32 c = (coss(angle) * 0x10000);
+    s32 s = (sins(angle) * 0x10000);
+    f32 *mtxp = (f32 *)mtx;
+    s32 i;
+    for (i = 0; i < 16; i++) {
+        *mtxp++ = 0;
     }
-    MATENTRY(0,  i)
-    MATENTRY(1,  j)
-    MATENTRY(4, -j)
-    MATENTRY(5,  i)
+    MATENTRY(0,  c)
+    MATENTRY(1,  s)
+    MATENTRY(4, -s)
+    MATENTRY(5,  c)
     ((s16 *) mtx)[10] = 1;
     ((s16 *) mtx)[15] = 1;
 }
 
+void create_transformation_from_matrices(Mat4 dst, Mat4 a1, Mat4 a2) {
+    f32 *dstp = (f32 *)dst;
+    f32 tx = a2[3][0];
+    f32 ty = a2[3][1];
+    f32 tz = a2[3][2];
+    f32 rx, ry, rz;
+    s32 i;
+    for (i = 0; i < 3; i++) {
+        rx = a2[i][0];
+        ry = a2[i][1];
+        rz = a2[i][2];
+        dstp[ 0] = (a1[0][0] * rx + a1[0][1] * ry + a1[0][2] * rz); //   vec3f_dot(a1[0], a2[i])
+        dstp[ 4] = (a1[1][0] * rx + a1[1][1] * ry + a1[1][2] * rz); //   vec3f_dot(a1[1], a2[i])
+        dstp[ 8] = (a1[2][0] * rx + a1[2][1] * ry + a1[2][2] * rz); //   vec3f_dot(a1[2], a2[i])
+        dstp[12] = (a1[3][0] * rx + a1[3][1] * ry + a1[3][2] * rz)  //   vec3f_dot(a1[3], a2[i])
+                 - (      tx * rx +       ty * ry +       tz * rz); // - vec3f_dot(a2[3], a2[i])
```

Comment:
more saved code for later?

Reply:
No, just an explanation of what it's doing.

Reply:
thats not much of an explanation, something like `dot product: a1[0] a2[0]` would make more sense